### PR TITLE
refactor: IO resource types, fix concurrent read/write and graceful close

### PR DESCRIPTION
**Pull request by piscisaureus**
TODO:
* add test for TLS graceful close
* also create separate type for `UnixStreamResource`.

### CODE REVIEW

**Comment by lucacasonato** [comment]
This should fix https://github.com/denoland/deno/issues/9032.

File: cli/tests/unit/tls_test.ts [hunk 1]
```diff
@@ -182,6 +183,149 @@ unitTest(
...
+unitTest(
+  { perms: { read: true, net: true } },
+  async function tlsServerStreamHalfClose(): Promise<void> {
+    const [serverConn, clientConn] = await tlsPair(3501);
+    await Promise.all([
+      sendCloseWrite(serverConn),
+      receiveCloseWrite(clientConn),
+    ]);
+  },
+);
```
Comment (ry) [inline cli/tests/unit/tls_test.ts:294]: Nice

File: cli/tests/unit/tls_test.ts [hunk 2]
```diff
@@ -182,6 +183,149 @@ unitTest(
   },
 );
 
+async function tlsPair(port: number): Promise<[Deno.Conn, Deno.Conn]> {
```
Comment (ry) [inline cli/tests/unit/tls_test.ts:186]: It's too bad we don't have a way to do ephemeral port allocaton

File: runtime/ops/io.rs [hunk 3]
```diff
@@ -142,6 +156,80 @@ pub struct FileMetadata {
...
+  }
+}
+
+#[derive(Debug)]
+pub struct ReadOnlyResource<S> {
+  stream: AsyncRefCell<S>,
+  cancel_handle: CancelHandle,
+}
+
+impl<S: 'static> From<S> for ReadOnlyResource<S> {
```
Comment (ry) [inline runtime/ops/io.rs:199]: These two From<S> -> ReadOnlyResource<S> seem backwards to me.... I would expect you to be able to go from


ReadOnlyResource<S> -> S

But not from any static lifetime object to ReadOnlyResource... am I missing something?

Comment (piscisaureus) [reply runtime/ops/io.rs:199]: I can add some type constraints (`S: AsyncRead + Unpin + 'static`), but these are not necessary.
Note that these type constraints are present on the `impl<S> ReadOnlyResource<S>` block.

File: runtime/ops/io.rs [hunk 4]
```diff
@@ -460,153 +530,163 @@ pub fn op_read(
...
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<TlsServerStreamResource>() {
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<UnixStreamResource>() {
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.write(&buf).await?
+  } else {
+    return Err(not_supported());
+  };
```
Comment (ry) [inline runtime/ops/io.rs:655]: I like that the conditional is now explicitly expressed here in the op. Much easier to read.

**Review by ry — APPROVED** [review]
Can't say I grok this fully yet - but it looks very clean and fixes bugs - so LGTM

**Pull request merged**